File: python/src/wdl_model/model/wdl_document.py
"""Root document node for parsed WDL source.

The WDL specification says that a document contains a version declaration followed by imports,
user-defined types, tasks, and workflows. This module mirrors that top-level structure and offers
typed views over the ordered element list.
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field
from typing import Dict

from wdl_model.model.base import WdlNode
from wdl_model.model.definitions import WdlEnum, WdlStruct, WdlTask, WdlWorkflow
from wdl_model.model.statements import WdlImport

from .wdl_version import WdlVersion


class WdlDocumentElement(WdlNode):
    """Marker for any top-level node that can appear directly in a WDL document."""

    pass


@dataclass
class WdlDocument(WdlNode):
    """Root node for a parsed WDL document."""

    wdlVersion: WdlVersion | None = None
    sourceLocation: str | None = None
    _elements: deque[WdlDocumentElement] = field(default_factory=deque)
    _importedDocuments: Dict[str, "WdlDocument"] = field(default_factory=dict)

    def getWdlVersion(self) -> WdlVersion | None:
        """Return the declared WDL version, if one was parsed."""
        return self.wdlVersion

    def setWdlVersion(self, wdlVersion: WdlVersion) -> None:
        """Set the declared WDL version."""
        self.wdlVersion = wdlVersion

    def getSourceLocation(self) -> str | None:
        """Return the source location identifier used for import resolution."""
        return self.sourceLocation

    def setSourceLocation(self, sourceLocation: str | None) -> None:
        """Set the source location identifier used for import resolution."""
        self.sourceLocation = sourceLocation

    def elements(self) -> deque[WdlDocumentElement]:
        """Return the ordered top-level elements exactly as they appeared in the source."""
        return self._elements

    def importStatements(self) -> list[WdlImport]:
        """Return only top-level import statements."""
        return [x for x in self._elements if isinstance(x, WdlImport)]

    def importedDocuments(self) -> Dict[str, "WdlDocument"]:
        """Return imported documents keyed by import identifier."""
        return self._importedDocuments

    def enums(self) -> list[WdlEnum]:
        """Return only top-level enum definitions."""
        return [x for x in self._elements if isinstance(x, WdlEnum)]

    def structs(self) -> list[WdlStruct]:
        """Return only top-level struct definitions."""
        return [x for x in self._elements if isinstance(x, WdlStruct)]

    def tasks(self) -> list[WdlTask]:
        """Return only top-level task definitions."""
        return [x for x in self._elements if isinstance(x, WdlTask)]

    def workflows(self) -> list[WdlWorkflow]:
        """Return only top-level workflow definitions."""
        return [x for x in self._elements if isinstance(x, WdlWorkflow)]

File: java/src/main/java/com/myriad/wdl/model/WdlDocument.java
package com.myriad.wdl.model;

import com.myriad.wdl.model.base.WdlNode;
import com.myriad.wdl.model.definitions.WdlEnum;
import com.myriad.wdl.model.definitions.WdlStruct;
import com.myriad.wdl.model.definitions.WdlTask;
import com.myriad.wdl.model.definitions.WdlWorkflow;
import com.myriad.wdl.model.statements.WdlImport;
import java.net.URI;
import java.util.ArrayDeque;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import lombok.Getter;
import lombok.Setter;

/**
 * Root node for a parsed WDL document.
 *
 * <p>The WDL specification says that a document contains a version declaration followed by imports,
 * user-defined types, tasks, and workflows. This class mirrors that shape and offers convenient
 * typed views over the underlying ordered element list.
 */
public final class WdlDocument implements WdlNode {

  /** Marker for any top-level element that can appear directly in a WDL document. */
  public interface WdlDocumentElement extends WdlNode {}

  @Getter @Setter private WdlVersion wdlVersion;
  @Getter @Setter private URI sourceLocation;
  private final ArrayDeque<WdlDocumentElement> elements = new ArrayDeque<>();
  private final Map<String, WdlDocument> importedDocuments = new LinkedHashMap<>();

  public WdlDocument() {}

  public WdlDocument(WdlVersion wdlVersion) {
    this.wdlVersion = wdlVersion;
  }

  /** Returns the ordered top-level elements exactly as they appeared in the source document. */
  public ArrayDeque<WdlDocumentElement> elements() {
    return elements;
  }

  /** Returns only top-level import statements. */
  public List<WdlImport> importStatements() {
    return elements.stream()
        .filter(WdlImport.class::isInstance)
        .map(WdlImport.class::cast)
        .collect(Collectors.toList());
  }

  /** Returns imported documents keyed by import identifier/location. */
  public Map<String, WdlDocument> importedDocuments() {
    return importedDocuments;
  }

  /** Returns only top-level enum definitions. */
  public List<WdlEnum> enums() {
    return elements.stream()
        .filter(WdlEnum.class::isInstance)
        .map(WdlEnum.class::cast)
        .collect(Collectors.toList());
  }

  /** Returns only top-level struct definitions. */
  public List<WdlStruct> structs() {
    return elements.stream()
        .filter(WdlStruct.class::isInstance)
        .map(WdlStruct.class::cast)
        .collect(Collectors.toList());
  }

  /** Returns only top-level task definitions. */
  public List<WdlTask> tasks() {
    return elements.stream()
        .filter(WdlTask.class::isInstance)
        .map(WdlTask.class::cast)
        .collect(Collectors.toList());
  }

  /** Returns only top-level workflow definitions. */
  public List<WdlWorkflow> workflows() {
    return elements.stream()
        .filter(WdlWorkflow.class::isInstance)
        .map(WdlWorkflow.class::cast)
        .collect(Collectors.toList());
  }

  @Override
  public String toString() {
    return getClass().getSimpleName();
  }
}
